### PR TITLE
Design for Same-Table-JOINs

### PR DESCRIPTION
**Pull request by acarbonetto**
### Description
Added documentation for a same-table join proposal. 

related: https://opensearch.org/docs/latest/field-types/join/
 
### Issues Resolved
* https://github.com/opensearch-project/sql/issues/1441
 
### Check List
- [ ] New functionality includes testing.
  - [ ] All tests pass, including unit test, integration test and doctest
- [x] New functionality has been documented.
  - [ ] New functionality has javadoc added
  - [x] New functionality has user manual doc added
- [x] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: docs/dev/opensearch-same-table-join-query.md [hunk 1]
```diff
@@ -0,0 +1,1349 @@
...
+    activate JoinTableContext
+    JoinTableContext->>FromClauseContext: Relation.attach(JoinRelation)
+    deactivate JoinTableContext
+    FromClauseContext->>AstBuilder: UnresolvedPlan
+    deactivate FromClauseContext
+```
+
+#### 6.2.3 Logical Plan Relation Visitor
+
+The following sequence diagram shows what the
```
Comment (matthewryanwells) [inline docs/dev/opensearch-same-table-join-query.md:1043]: Was this sentence cut off?

File: docs/dev/opensearch-same-table-join-query.md [hunk 2]
```diff
@@ -0,0 +1,1349 @@
...
+    participant DataSourceService
+    participant Table
+    participant TypeEnvironment
+
+    SQLService->>Analyzer: analyze
+    Analyzer->>Relation: visitRelation
+    Relation->>DataSourceService: getTable
+    DataSourceService-->>Relation: Table
+    Relation->>Table: getFieldTypes
+    Table->>TypeEnvironment: define
```
Comment (Yury-Fridlyand) [inline docs/dev/opensearch-same-table-join-query.md:1060]: You are missing returns from `TypeEnvironment` and `Table` back to `Relation`

File: docs/dev/opensearch-same-table-join-query.md [hunk 3]
```diff
@@ -0,0 +1,1349 @@
...
+    Relation->>Table: getReservedFieldTypes
+    Table->>TypeEnvironment: define
+    Relation->>TypeEnvironment: define(node.alias)
+    Relation->>Analyzer: LogicalRelation
+    Analyzer->>JoinRelation: visitJoinRelation
+    JoinRelation->>TypeEnvironment: getOpenSearchJoinType
+    TypeEnvironment->>JoinRelation: OpenSearchJoinType
+    JoinRelation->>JoinRelation: setJoinRelationContext
+    JoinRelation->>Analyzer: <<LogicalRelation>>
```
Comment (Yury-Fridlyand) [inline docs/dev/opensearch-same-table-join-query.md:1069]:
```suggestion
    JoinRelation->>Analyzer: LogicalRelation
```

File: docs/dev/opensearch-same-table-join-query.md [hunk 4]
```diff
@@ -0,0 +1,1349 @@
...
+## 4. Use Cases
+
+Use Cases with examples for all the ways the join field can be accessed in OpenSearch. Each use case contains examples, 
+with data and mappings defined in the Annex.  
+
+### 4.1 Case SELECT... FROM... JOIN ON parent relation
+
+Summary:
+
+Example query (SQL):
```
[the remark this answers is not in the document]
Comment (Yury-Fridlyand) [reply docs/dev/opensearch-same-table-join-query.md:74]: I found it on the very bottom. Please add a link here to the data.

File: docs/dev/opensearch-same-table-join-query.md [hunk 5]
```diff
@@ -0,0 +1,1349 @@
...
+        }
+      }
+    }
+  }
+}
+```
+
+## Annex B - Data for examples
+
+The following section includes data that has been formatted to be used by the OpenSearch `got/_bulk` ingest endpoint.
```
Comment (Yury-Fridlyand) [inline docs/dev/opensearch-same-table-join-query.md:1294]: Mention that all data to be added by single index named `got` (not multiple indices)

**Comment by Swiddis** [comment]
Is this still in progress? Should we close it?

**Comment by Swiddis** [comment]
Closing as stale -- feel free to reopen if work is resumed

**Pull request closed without merging**